In [1]:
import sys
import os
from tempfile import NamedTemporaryFile
from gc import collect

sys.path.append('../util')
from meters import ThroughputMeter
import ncgen

import netCDF4
import numpy as np
import pylab as pl

os.environ['TMPDIR'] = os.getenv('HOME') + '/tmp/'

In [4]:
results = []
for timedimpos in [True, False]:
    for num_vars in range(1, 20, 2):
        print("Creating a time-{} NetCDF file with {} variables".format('major' if timedimpos else 'minor', num_vars))
        with NamedTemporaryFile(suffix='.nc', delete=False) as f:
            nc = ncgen.make_multivariable_nc(f.name, num_vars=num_vars, unlim=True)
            nc.close()
        nc = netCDF4.Dataset(f.name, 'r')
        with ThroughputMeter() as t:
            a = nc.variables['var_0'][:, :, :]
        
        # python-netCDF4 seems to leak file descriptors
        # We have to take a lot of steps to make sure that the files get closed and that
        # the space gets reclaimed by the OS
        nc.close
        del nc
        print("Removing {}".format(f.name))
        os.remove(f.name)
        f.close()
        collect()
        results.append((timedimpos, num_vars, t.megabytes_per_second(a)))

Creating a time-major NetCDF file with 1 variables
Removing /tmp/tmpc8hwsjyg.nc
353.100 MB in 0.325 seconds at 1087.240 MB / sec
Creating a time-major NetCDF file with 3 variables
Removing /tmp/tmpna0b06_q.nc
353.100 MB in 1.61 seconds at 218.934 MB / sec
Creating a time-major NetCDF file with 5 variables
Removing /tmp/tmptt4wgo66.nc
353.100 MB in 1.37 seconds at 257.676 MB / sec
Creating a time-major NetCDF file with 7 variables
Removing /tmp/tmpqei0kdbn.nc
353.100 MB in 1.54 seconds at 229.723 MB / sec
Creating a time-major NetCDF file with 9 variables
Removing /tmp/tmpdqgj2rl8.nc
353.100 MB in 1.34 seconds at 263.825 MB / sec
Creating a time-major NetCDF file with 11 variables
Removing /tmp/tmppedeqp3p.nc
353.100 MB in 1.4 seconds at 251.935 MB / sec
Creating a time-major NetCDF file with 13 variables
Removing /tmp/tmphk5w20gc.nc
353.100 MB in 1.43 seconds at 246.689 MB / sec
Creating a time-major NetCDF file with 15 variables
Removing /tmp/tmpnj57_yn7.nc
353.100 MB in 1.34 seconds 

In [5]:
results

[(True, 1, 1087.2404125473233),
 (True, 3, 218.93397865071705),
 (True, 5, 257.67642439461565),
 (True, 7, 229.72277825750632),
 (True, 9, 263.82488242838821),
 (True, 11, 251.9347266053997),
 (True, 13, 246.68876693873921),
 (True, 15, 263.05926637202754),
 (True, 17, 250.22388421568212),
 (True, 19, 263.23329214520027),
 (False, 1, 864.32366279989287),
 (False, 3, 234.00696139599549),
 (False, 5, 229.55866364971351),
 (False, 7, 253.4975747540264),
 (False, 9, 255.85299426847223),
 (False, 11, 251.72965287486244),
 (False, 13, 232.15028089887642),
 (False, 15, 256.18519139325821),
 (False, 17, 251.70158775155801),
 (False, 19, 251.96404400791309)]

and again (run from the command line)...

Creating a time-major NetCDF file with 1 variables
Removing /tmp/tmpfbrgznqt.nc
353.100 MB in 0.314 seconds at 1126.189 MB / sec
Creating a time-major NetCDF file with 3 variables
Removing /tmp/tmpz618vjky.nc
353.100 MB in 1.44 seconds at 245.712 MB / sec
Creating a time-major NetCDF file with 5 variables
Removing /tmp/tmp0n7lzuqo.nc
353.100 MB in 1.38 seconds at 256.594 MB / sec
Creating a time-major NetCDF file with 7 variables
Removing /tmp/tmph_hqkf46.nc
353.100 MB in 1.68 seconds at 210.260 MB / sec
Creating a time-major NetCDF file with 9 variables
Removing /tmp/tmp93eqp9x_.nc
353.100 MB in 1.67 seconds at 211.540 MB / sec
Creating a time-major NetCDF file with 11 variables
Removing /tmp/tmpwjrusbht.nc
353.100 MB in 1.74 seconds at 203.277 MB / sec
Creating a time-major NetCDF file with 13 variables
Removing /tmp/tmpqbkvc90n.nc
353.100 MB in 1.61 seconds at 219.115 MB / sec
Creating a time-major NetCDF file with 15 variables
Removing /tmp/tmp4d2ui9st.nc
353.100 MB in 1.55 seconds at 227.796 MB / sec
Creating a time-major NetCDF file with 17 variables
Removing /tmp/tmp7v37u_hq.nc
353.100 MB in 1.71 seconds at 206.714 MB / sec
Creating a time-major NetCDF file with 19 variables
Removing /tmp/tmp3z4fz63e.nc
353.100 MB in 1.7 seconds at 207.965 MB / sec
Creating a time-minor NetCDF file with 1 variables
Removing /tmp/tmp7dlf754y.nc
353.100 MB in 2.29 seconds at 154.183 MB / sec
Creating a time-minor NetCDF file with 3 variables
Removing /tmp/tmp17wjppn9.nc
353.100 MB in 1.1 seconds at 319.745 MB / sec
Creating a time-minor NetCDF file with 5 variables
Removing /tmp/tmpwz7etvrr.nc
353.100 MB in 0.647 seconds at 545.571 MB / sec
Creating a time-minor NetCDF file with 7 variables
Removing /tmp/tmpxlslqyjd.nc
353.100 MB in 1.72 seconds at 205.496 MB / sec
Creating a time-minor NetCDF file with 9 variables
Removing /tmp/tmpw9h_8zyi.nc
353.100 MB in 1.64 seconds at 214.965 MB / sec
Creating a time-minor NetCDF file with 11 variables
Removing /tmp/tmpjp67_xg0.nc
353.100 MB in 1.67 seconds at 211.767 MB / sec
Creating a time-minor NetCDF file with 13 variables
Removing /tmp/tmp770n9j5y.nc
353.100 MB in 1.71 seconds at 206.908 MB / sec
Creating a time-minor NetCDF file with 15 variables
Removing /tmp/tmpyjvuaksu.nc
353.100 MB in 1.7 seconds at 208.188 MB / sec
Creating a time-minor NetCDF file with 17 variables
Removing /tmp/tmpcf604xq5.nc
353.100 MB in 1.63 seconds at 216.240 MB / sec
Creating a time-minor NetCDF file with 19 variables
Removing /tmp/tmpw5r4ictr.nc
353.100 MB in 1.77 seconds at 198.934 MB / sec

There appears to be no significant difference between the read speed of time-major vs. time-minor data at this scale (using the bc_400m grid).